### PR TITLE
New Validation System

### PR DESCRIPTION
**Pull request by tlouisse**
Please only merge (to a 'next' branch) when changes for all things relying on this are updated...

## Impact of all changes

**A ) MIGRATION LIST**
```js
/**
 * 1a. Imports 'from'
 */

import {
  isString, equalsLength, minLength, maxLength, minMaxLength, isEmail,
  isStringValidator, equalsLengthValidator, minLengthValidator, maxLengthValidator, minMaxLengthValidator, isEmailValidator,
  
  isNumber, minNumber, maxNumber, minMaxNumber, 
  isNumberValidator, minNumberValidator, maxNumberValidator, minMaxNumberValidator,
  
  isDate, minDate, maxDate, isDateDisabled, minMaxDate, 
  isDateValidator, minDateValidator, maxDateValidator, minMaxDateValidator, isDateDisabledValidator,
  
  randomOk, defaultOk,
  randomOkValidator, defaultOkValidator,
} from '@lion/validate'; // '@lion/validate/src/validators.js'


/**
 * 1b. Imports 'to'
 */

import {
  IsString, EqualsLength, MinLength, MaxLength, MinMaxLength, IsEmail,
  
  IsNumber, MinNumber, MaxNumber, MinMaxNumber, 
  
  IsDate, MinDate, MaxDate, IsDateDisabled, MinMaxDate, 
  
  DefaultSuccess,
} from '@lion/validate';



/**
 * 2a. Templates 'from'
 */

const paramRef = new Date();
const cfgRef = { con: 'fig' };

const tplFrom = html`
<lion-input 
  .errorValidators="${[minLengthValidator(1, { x: 'asd' }), ]}" 
  .warningValidators="${[isNumberValidator()]}"
  .infoValidators="${[maxDateValidator(paramRef, cfgRef)]}"
  .successValidators="${[defaultOk()]}">
</lion-input>
`;



/** 
 * 2b. Templates 'to'
 */

const tplTo = html`
<lion-input 
  .validators="${[
    new MinLength(1, { x: 'asd' }), 
    new IsNumber(null, { type: 'warning'}),
    new MaxDate(paramRef, { ...cfgRef, type: 'info' }),
    new DefaultSuccess(),
  ]}">
</lion-input>
`;



/** 
 * 3a. Custom Validator definitions 'from'
 */

export function isIban(...args) {
  return {
    'my-app-isIban': /\d{5}([ \-]\d{4})?$/.test(args[0]),
  };
}

export const isIbanValidator = (...factoryParams) => [
  (...params) => ({ 'my-app-isIban': isIban(...params) }),
  ...factoryParams,
];

localize.loadNamespace({
  'lion-validate+my-app-isIban': locale => {
    return import(`./translations/${locale}.js`);
  },
});


/**
 * 3b. Custom Validator definitions 'to'
 */

import { Validator } from '@lion/validate';

const nsMyAppIsIban = localize.loadNamespace({
  'lion-validate+my-app-isIban': locale => {
    return import(`./translations/${locale}.js`);
  },
});


export class IsIban extends Validator {
  constructor(...args) {
    super(...args);
    this.name = 'IsIban';
  }
  
  execute(...args) {
  	return !isIban(...args)['my-app-isIban'];
  }
  
  static async getMessage(data) {
  	await nsMyAppIsIban;
    return localize.msg('lion-validate+my-app-isIban:error.my-app-isIban', await transformData(data));
  }
}

// TODO: Name 'IsIban' needs to be backwards compatible with eventually used:
// - .error["my-app-isIban"]
// - All references of isIbanValidator need to be replaced inside .{x}Validators
// - 

/** 
 * 4a. Error states (from)
 */


myInputRef.errorState;
myInputRef.error.required;

/** 
 * 4a. Error states (to)
 */
myInputRef.hasError;
myInputRef.errorStates.Required;



/** 
 * 5a. Configurations of default validators inside components (from)
 */
class MyComp extends HTMLElement {

  getValidatorsForType(type) {
    if (type === 'error') {
      return [isDateValidator()].concat(super.getValidatorsForType(type) || []);
    }
    else if (type === 'warning') {
      return [equalsLengthValidator(2)].concat(super.getValidatorsForType(type) || []);
    }
    return super.getValidatorsForType(type);
  }
}

/** 
 * 5b. Configurations of default validators inside components (to)
 */

class MyCompStill extends HTMLElement {
  constructor() {
    super();
    this.defaultValidators.push(new IsDate());
    this.defaultValidators.push(new EqualsLength(2, { type: 'warning' }));
  }
}


/** 
 * 6a. getFieldName (from)
 */

const tplFieldFrom = html`
<lion-input .getFieldName="${() => 'xyz'}"></lion-input>
`;


/** 
 * 6b. .fieldName (to)
 */

const tplFieldTo = html`
<lion-input .fieldName="${'xyz'}"></lion-input>
`;


/** 
 * 7a. show{Type}Condition (from)
 */

class MyCompX extends HTMLElement {
  showErrorCondition(...args) {
  	const parent = super.showErrorCondition(...args);
    return parent && (1 + 1 === 3);
  }
}



/** 
 * 7b. show{Type}Condition (to)
 */
class MyCompXStill extends HTMLElement {
  _prioritizeAndFilterFeedback({ validationResult }) {
    const parent = super._prioritizeAndFilterFeedback({ validationResult });
    return parent && (1 + 1 === 3);
  }
}
```

**B ) COMPONENTS AFFECTED** 
- lion-field
- lion-input
- lion-textarea
- lion-select
- lion-select-rich
- lion-fieldset
- lion-choice-input/ radio/checkbox/option (?)
- lion-radiogroup (?)
- lion-checkbox-group  (?)
- lion-validate (all docs)
- lion-input-amount / ing-input-amount 
- lion-input-email / ing-input-email
- lion-input-iban / ing-input-iban
- lion-input-date / ing-input-date
- lion-input-datepicker / ing-input-datepicker
- lion-form
- lion-form-system -> demos

- See lion-input-date for an example that is already migrated

### CODE REVIEW

File: packages/input/stories/validation-string.stories.js [hunk 1]
```diff
@@ -1,131 +1,189 @@
...
+        label="Async validation"
+        .validators="${[new AsyncValidator()]}"
+        .modelValue="${'123'}"
+      ></lion-input>
+    `;
+  })
+  .add('Dynamic parameter changes', () => {
+    const beginDate = new Date('09/09/1990');
+    const minDateValidatorRef = new MinDate(beginDate, {
+      message: 'Fill in a date after your birth date',
```
Comment (tlouisse) [inline packages/input/stories/validation-string.stories.js]: rewrite to getMessage()

File: packages/validate/src/ResultValidator.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import { Validator } from './Validator.js';
+
+/**
+ * @desc Instead of evaluating the result of a regular validator, a HolisticValidator looks
```
Comment (tlouisse) [inline packages/validate/src/ResultValidator.js]: rename to ResultValidator

File: packages/validate/src/ValidateMixin.js [hunk 3]
```diff
@@ -1,624 +1,399 @@
...
-
-        if (this._feedbackElement) {
-          // Only write to light DOM not put there by Application Developer, but by <lion-component>
-          if (typeof this._feedbackElement.renderFeedback === 'function') {
-            this._feedbackElement.renderFeedback(this.getValidationStates(), this.message, details);
-          } else {
-            this.renderFeedback(this.getValidationStates(), this.message, details);
+          if (c.has(`${type}States`)) {
+            this.dispatchEvent(
+              new Event(`${type}-states-changed`, { bubbles: true, composed: true }),
```
Comment (daKmoR) [inline packages/validate/src/ValidateMixin.js]: please no `composed: true` as this make events basically global...

File: packages/validate/src/ValidateMixin.js [hunk 4]
```diff
@@ -1,663 +1,616 @@
...
-          let validatorFn = validatorArray[0];
-          const validatorParams = validatorArray[1];
-          const validatorConfig = validatorArray[2];
-
-          let isRequiredValidator = false; // Whether the current is the required validator
-          if (typeof validatorFn === 'string' && validatorFn === 'required' && this.__isRequired) {
-            validatorFn = this.__isRequired;
-            isRequiredValidator = true;
-          }
+      async _updateFeedbackComponent() {
```
Comment (daKmoR) [inline packages/validate/src/ValidateMixin.js]:
```js
// TODO: as _updateFeedbackComponent is called without await in updated and in an event it means that a getting
// the messageMap might take longer and a call afterwards who set's the value to nothing gets overwritten by the
// later resolve of the messageMap
// therefore we await here for an existing feedbackComplete however this is not ideal as it means we wait
// for messages which we then never display => use a que? handle promise/resolve/reject manually without await?
// feedbackNode should handle promises? maybe not
if (this.feedbackComplete) {
  await this.feedbackComplete;
}
```

=> 3 failed tests as we probably would now need to await 2 times in the tests... 
=> maybe already invest in a proper solution?
=> or do "above" hack and skip tests or "hack" them as well... and then release so we can move forward... and in parallel start working on this

**Pull request merged**